### ISSUE TITLE
Time Interpolation: Mass-Conserving Interpolation for Accumulated Fields

### ISSUE THREAD

**Issue by Rilwan-Adewoyin** [ecmwf/anemoi-core#402]
### Is your feature request related to a problem? Please describe.

<h1><strong>Mass-Conserving Interpolation for Accumulated Fields</strong></h1>

<p>When down-scaling <strong>accumulated</strong> variables—such as total precipitation (<code inline="">tp</code>) and convective precipitation (<code inline="">cp</code>)—from a 6-hourly forecast to hourly resolution, naïve interpolation yields hourly values that <strong>do not sum to the original 6-hour total</strong>.<br>
This breaks water-balance conservation and can distort any downstream verification or post-processing steps that integrate precipitation over time.</p>

### Proposed Algorithm

1. **Predict per-hour logits**
   Generate one raw score $z_{t+k}$ for each hourly sub-interval, $k = 1,\dots,6$.

2. **Softmax to obtain allocation weights**

   $$w_{t+k} = \frac{\exp(z_{t+k})}{\sum_{j=1}^{6}\exp(z_{t+j})}, \quad k = 1,\dots,6.$$

3. **Scale by the 6-hour accumulation**

   $$\Delta A_{t+k} = w_{t+k}\, A_{t+6}, \quad k = 1,\dots,6.$$

### Properties

* **Mass-conserving**:

  $$\sum_{k=1}^{6}\Delta A_{t+k} = A_{t+6}$$

  (since softmax ensures $\sum_{k=1}^{6} w_{t+k} = 1$).

* **Non-negative**: Each hourly increment is guaranteed non-negative, $\Delta A_{t+k}\ge 0$.

* **Differentiable**: The entire interpolation procedure is end-to-end differentiable, allowing gradient-based optimization.

### Describe alternatives you've considered

Any alternative implementations for the 1st and 2nd point?

**Comment by radiradev** [comment]
Very cool idea! @frazane also wanted to explore this for precipitation. 

One issue is that currently we do:
```
  explicit_times:
    input: [0,6]
    target: [1, 2, 3, 4, 5]
```
So we only have 5 targets, but you'd need 6 for an accumulated field to be able to spread it evenly.

**Comment by clessig** [comment]
What happens if the original dataset has hourly accumulated data, we train on 6-hourly data (properly accumulated), and then want to train the time interpolator on the hourly data, which is not necessarily the accumulated precip uniformly sparead in time?

**Comment by Rilwan-Adewoyin** [comment]
Thanks for raising that complication @radiradev , 

@frazane the fixed 0 logit for the final term sounds like a very slick way to implement this [ Happy for you to update the initial post with your improved suggestion ] The main motivation from us was also for precip. Did @frazane  or @radiradev have plans / want to tackle this PR?

**Comment by Rilwan-Adewoyin** [comment]
> What happens if the original dataset has hourly accumulated data, we train on 6-hourly data (properly accumulated), and then want to train the time interpolator on the hourly data, which is not necessarily the accumulated precip uniformly sparead in time?

@clessig. :  
Okay so if I understand your scenario: AIFS train 6hrly accum from 1hrly data -> Interpolator train on the 1hrly accum from 1hrly data. 

To handle this case _I think_ there is a specific setup for the config.data file that allows anemoi-datasets to accumulate 1hrly accum fields into 6hrly fields upon reading the file in. This would handle the 6hrly input accum. The second part of the solution to this - ensuring we can target the original 1hrly fields would require the multi-dataset loading logic to be completed #279 so the 6hrly accumulated field and the 1hrly accumulated can both be loaded in in tandem

Alternatively @mchantry did discuss more on-the-fly methods to handle this (such as some bespoke time interpolator code logic to handle this scenario) which could also be explored/discussed a quicker solution

**Comment by clessig** [comment]
> > What happens if the original dataset has hourly accumulated data, we train on 6-hourly data (properly accumulated), and then want to train the time interpolator on the hourly data, which is not necessarily the accumulated precip uniformly sparead in time?
> 
> [@clessig](https://github.com/clessig). : Okay so if I understand your scenario: AIFS train on 6hrly accum from 1hrly data -> Interpolator train on the 1hrly accum from 1hrly data. -> How do we train interpolator on non AIFS data (that isn't 6hrly accums), but still allow it to be useable on AIFS models that do output 6hrly accums
> 
> To handle this case _I think_ there is a specific setup for the config.data file that allows anemoi-datasets to accumulate 1hrly accum fields into 6hrly fields upon reading the file in. This would handle the 6hrly input accum. The second part of the solution to this - ensuring we can target the original 1hrly fields would require the multi-dataset loading logic to be completed [#279](https://github.com/ecmwf/anemoi-core/discussions/279) so the 6hrly accumulated field and the 1hrly accumulated can both be loaded in in tandem
> 
> Alternatively [@mchantry](https://github.com/mchantry) did discuss more on-the-fly methods to handle this (such as some bespoke time interpolator code logic to handle this scenario) which could also be explored/discussed as quicker solution

I just wanted to flag this case so that it is not assumed that the precip is uniformly distributed in the 6h (or whatever the coarser temporal resolution).

**Comment by Rilwan-Adewoyin** [comment]
Update: I've opened a branch [402-time-interpolation...](https://github.com/ecmwf/anemoi-core/tree/feat%2F402-time-interpolation-mass-conserving-interpolation-for-accumulated-fields ) - which contains an implementation of:

1. The loss I suggested in the opening post as well as the improvement suggested by @frazane 
2. Configurability to handle the different required behaviours for treating accumulation (mentioned in @clessig previous message )

Current Status:
- The ability to properly train with m hourly accum fields and predict n hourly accum fields where n<m, requires a user to pass in multiple datasets. As such this functionality is pending the parallel work on datasets #279 . This is also the only scenario where mass conservation can be used.

**Comment by Rilwan-Adewoyin** [comment]
### Follow-up: Mass-conserving interpolation for accumulated fields (dataset configs & boundary conditions)

Following up from [[#402](https://github.com/ecmwf/anemoi-core/issues/402)](https://github.com/ecmwf/anemoi-core/issues/402) and based on suggestions from @Magnus and @Matt, we’ve refined the approach for handling accumulated fields when interpolating to higher-resolution intervals.

#### Key points

* ✅ **Custom dataset configs** can be used to achieve simultaneous loading of:

  * accumulated boundary conditions for accumulated variables (e.g. 6-hourly totals), and
  * interval targets accumulated over a higher temporal resolution (e.g. 1-hourly totals).

* ⚙️ The method relies on defining:

  * the accumulated input boundary conditions as **forcings** in the config,
  * the higher-resolution interval targets as **diagnostics**.

* 📘 **Example:** interpolating **6-hour `tp` accumulations → 1-hour `tp` accumulations**:

  1. Rename the 6-hour field to `tp_accum`
  2. Define `tp_accum` as a **forcing**
  3. Define `tp` (1-hour field) as a **diagnostic**

* 🔧 For accumulated variables with **two boundary conditions** (e.g. `tp_accum` at `t=0` and `t=6`):

  * we always set the **left boundary (`t=0`) to 0**.
  * This matches inference scenarios, where operational forecast models typically initialise accumulated fields at zero at step 0.

* 📎 More details on this design are available in this PR:
[Anemoi-core: Associated PR](https://github.com/ecmwf/anemoi-core/pull/544)
  [anemoi-transform: feat/acum\_to\_interval](https://github.com/ecmwf/anemoi-transform/tree/feat/acum_to_interval)

---

#### Contributions & TODOs

@sahahner  has proposed the improvement of my initial mass-conservation logic to better align with the **postprocessors** framework:

1. 💡 **DONE** Extended my initial-postprocessor that added a zero at index `0` of a sequence so that the index is **configurable by the user** - index 0 is overwritten can be configured.
2. 📌 **TODO:** Explore moving the mass conservation logic into a postprocessor.

   * This requires discussion — it’s not straightforward because postprocessors currently run **after every forward pass**, acting only on that single output.
   * For interpolation across **N timesteps**, mass conservation would need a postprocessor that runs **only after the final (`Nth`) step**, with access to the **outputs from all N steps**.
   * Some restructuring of the postprocessing pipeline may be required to support this.
